core/embed/rust/src/trezorhal/uzlib.rs

```diff
@@ -0,0 +1,42 @@
+use super::ffi;
+use core::ptr;
+
+pub const UZLIB_WINDOW_SIZE: usize = 1 << 10;
+pub use ffi::uzlib_uncomp;
+
+impl Default for ffi::uzlib_uncomp {
+    fn default() -> Self {
+        let mut s = ::core::mem::MaybeUninit::<Self>::uninit();
+        unsafe {
+            ptr::write_bytes(s.as_mut_ptr(), 0, 1);
+            s.assume_init()
+        }
+    }
+}
+
+pub fn uzlib_prepare(
+    decomp: &mut uzlib_uncomp,
+    window: Option<&mut [u8]>,
+    src: &[u8],
+    dest: &mut [u8],
+) {
+    unsafe {
+        decomp.source = src.as_ptr();
+        decomp.source_limit = src.as_ptr().add(src.len());
+        decomp.dest = dest.as_mut_ptr();
+        decomp.dest_limit = dest.as_mut_ptr().add(dest.len());
+
+        if let Some(w) = window {
+            ffi::uzlib_uncompress_init(decomp, w.as_ptr() as _, w.len() as u32);
+        } else {
+            ffi::uzlib_uncompress_init(decomp, ptr::null_mut(), 0);
+        }
+    }
+}
+
+pub fn uzlib_uncompress(decomp: &mut uzlib_uncomp, dest: &mut [u8]) -> i32 {
```

Comment:
well yes, the fastest is to do whole decompression in one call (=dest buffer is the size of data), like its done in rect_rounded2_partial function exactly for the performance reasons, but for some functionality this is not usable (large icons/images would take too much ram). in rect_rounded2_partial the size of icon is limited to 64*64, so the size is not that critical. If needed (and possible for the given algorithm), we can try using the buffer size somewhere in the middle. What exactly would you like faster?

Reply:
In the past I was unhappy with homescreen decompression speed. That is precisely the case where it won't fit into RAM as a whole, but scanning in chunks could presumably bring speedups.
Also the small black&white icons could certainly be decompressed in one go, RAM should not be an issue there and I see no reason not to do this small optimization.

Reply:
ok i will make some experiments

Reply:
Some results from the testing:
- the fastest is decompress as a whole, without window usage
- slowest is to decompress byte by byte
This is where intuitive results end though.
When decompressing by chunks, fastest seem to be using very small chunks of 2 or 4 bytes, after that, with increasing chunk size the speed actually goes down. I was testing this on two images, the best chunk size differ in each. I also tried comiling the uzlib with -O3 for one image, which sped up the decompressing quite nicely, but didn't change the dependence on the chunk size.
Even when decompressing in one chunk when using window, its way slower than without window usage or small chunks.

Just to have some idea what increases we can get:
Decompressing by best chunk was 17% to 65% faster than per byte
Decompressing by 128 byte chunk was 7% to 40% faster than per byte.
Decompressing whole image without window was 70% to 110% faster than per byte.
This will differ a lot by image. Maybe testing on more images would bring better results. 

So based on this, i would suggest using chunks of 4 bytes for decompression of large images (that means modify display_icon, display_image functions) and implementing new display_small_icon with some reasonable limits (loader using 64*64 so perhaps that) where we could do the whole decompression at once.

Reply:
Curious! Nice research.

I read a bit of the uzlib code. The no-window gains are obvious; essentially, whenever you're using a window, you are copying every byte twice, plus some conditional handling. I suspect we would gain another 2x speedup just by splittting into `uzlib_uncompress_window` and `uzlib_uncompress_nowindow` (honestly that's one of the things `-O3` is probably doing behind the scenes)

The with-window gains are super questionable, because internally the library decodes byte by byte anyway. However, if empirical research tells us that 4-byte chunks are faster, let's use that. (something something CPU word size??)

I'm also wondering if we could hide this inside `display_icon`, by looking at how big we expect the result to be, and choosing the decompression mode? Instead of having a separate `display_small_icon`.

Reply:
...actually, I don't think `display_icon` is ever used for large images. Let's place a hard limit on the icon size and always decompress the full thing; and keep the chunked variant only for `display_image`

Reply:
Largest seems to be 'homescreen_model_r' with 92*92, which is some 4kB of data, with current impl stored on stack - that seems already too much. But apart from this one it looks like we should fit into 64*64 (2kb) which we are already decompressing full in display_loader. 

We would need to convert the homescreen to image format, which will produce larger data stored on flash as it uses full colors (2 byte per pix instead of 4 bits), and also allow display_image function on model R, which we didn't need so far.

The hiding large/small in one function could work, but i imagine we would need to take the stack space for full decompression of smaller icons even when decompressing by small chunks. But that doesn't seem like a big problem.

Also looking at the display_image function: it's already decompressing by chunks of 2 bytes unlike display_icon, so there is probably not that much to gain.

Reply:
aren't icons 1bpp so 64x64 should come out to 512 bytes? (and the homescreen is 1058 bytes)

Reply:
no they are greyscale 4bpp

Reply:
Right.
So, there's the TR homescreen, and we really don't want to upscale that to fullcolor.
And there is the loader icon which is 64x64 (but will it be calling `display_icon`? the decompression is internal)
Other icons are at most 24x24, so that is IMHO the right size limit.

The fact that TR homescreen is rendered via `display_icon` is semantically incorrect, but i don't see anything elegant to do with that for now. We _could_ redefine `display_image` so that on TR it accepts 4bpp images, or we could stuff the feature into `display_icon`, or create a separate function for that purpose -- any other options?

Reply:
We could also implement a B&W icon/image function, modify the data generation to use 1bpp and use larger image size limit there. 

Though i am not sure that setting a hard limit on either of these functions is that great and might need reevaluation when designers desire slightly larger for some reason.

There probably also isn't reason to have the limit strictly squared, using just limit on pixel count will give us a bit more manipulation space.

- loader doesn't call display_icon, i just used that as reference of what we are using elsewhere to estimate what we can(need to be able to) afford on stack
- model 1 homescreen also doesn't fit 24*24, didn't check others so can't confirm nor deny that the rest fits

Anyway, this feels like it should be maybe moved to a separate issue

Reply:
agreed: https://github.com/trezor/trezor-firmware/issues/2414